feat(sdk): Spaces

stefanceriu opened the pull request:
This series of patches introduces a top level UI oriented interface for interacting with [spaces](https://spec.matrix.org/latest/client-server-api/#spaces). 

The `SpaceService` offers:
* a method for retrieving one's top level `joined_spaces` and a `subscribe_to_joined_spaces` stream that automatically updates based on client activity
    * to do so the service builds a graph from all the `m.space.parent` and `m.space.children` state events it knows about, removing cycles and then only keeping parent spaces
* a `SpaceRoomList` that can be used to interact with the [rooms hierarchy](https://spec.matrix.org/latest/client-server-api/#get_matrixclientv1roomsroomidhierarchy) endpoint and retrieve rooms referencing a certain parent
    * the room list is responsible for holding the pagination state as well as listening to client activity and automatically publishing room state updates through its `subscribe_to_room_updates` publisher (e.g. when joining one of the rooms)

I do appreciate this is a pretty chunky PR but it should be straight forward to review commit by commit.

pixlwave commented on bindings/matrix-sdk-ffi/src/spaces.rs:
```diff
@@ -0,0 +1,180 @@
+        self.inner.pagination_state().into()
+    }
+
+    pub fn subscribe_to_pagiation_state_updates(
```
```suggestion
    pub fn subscribe_to_pagination_state_updates(
```

pixlwave replied:
Resolved but remains unfixed 😢

Hywan reviewed (commented):
First fast review. Keep going!

pixlwave commented on bindings/matrix-sdk-ffi/src/client.rs:
```diff
@@ -1259,6 +1261,11 @@ impl Client {
         SyncServiceBuilder::new((*self.inner).clone(), self.utd_hook_manager.get().cloned())
     }
 
+    pub fn spaces_service(&self) -> Arc<SpaceService> {
```
Minor nit: This would be more consistent with the types.
```suggestion
    pub fn space_service(&self) -> Arc<SpaceService> {
```

pixlwave commented on bindings/matrix-sdk-ffi/src/spaces.rs:
```diff
@@ -0,0 +1,214 @@
+}
+
+impl SpaceService {
+    pub fn new(inner: UISpaceService) -> Self {
```
Does this one need to be `pub`?

pixlwave commented on bindings/matrix-sdk-ffi/src/spaces.rs:
```diff
@@ -0,0 +1,214 @@
+}
+
+impl SpaceServiceRoomList {
+    pub fn new(inner: UISpaceServiceRoomList) -> Self {
```
Same here?

pixlwave commented on bindings/matrix-sdk-ffi/src/spaces.rs:
```diff
@@ -0,0 +1,214 @@
+    Loading,
+}
+
+impl From<UISpaceServiceRoomListPaginationState> for SpaceServiceRoomListPaginationState {
+    fn from(state: UISpaceServiceRoomListPaginationState) -> Self {
+        match state {
+            UISpaceServiceRoomListPaginationState::Idle { end_reached } => {
+                SpaceServiceRoomListPaginationState::Idle { end_reached }
+            }
+            UISpaceServiceRoomListPaginationState::Loading => {
+                SpaceServiceRoomListPaginationState::Loading
+            }
+        }
+    }
+}
```
Is there a reason not to put the `#[derive(uniffi::Enum)]` on the enum in the UI crate and remove the FFI's copy entirely?

pixlwave commented on bindings/matrix-sdk-ffi/src/spaces.rs:
```diff
@@ -0,0 +1,214 @@
+    fn on_update(&self, pagination_state: SpaceServiceRoomListPaginationState);
+}
+
+#[matrix_sdk_ffi_macros::export(callback_interface)]
+pub trait SpaceServiceRoomListEntriesListener: SendOutsideWasm + SyncOutsideWasm + Debug {
+    fn on_update(&self, rooms: Vec<SpaceServiceRoom>);
+}
+
+#[matrix_sdk_ffi_macros::export(callback_interface)]
+pub trait SpaceServiceJoinedSpacesListener: SendOutsideWasm + SyncOutsideWasm + Debug {
+    fn on_update(&self, rooms: Vec<SpaceServiceRoom>);
+}
```
Up to you but I noticed whilst conforming out `SDKListener` to these that they're identical. So maybe a single generic `SpaceServiceSpaceRoomsListener` could suffice here 🌵

pixlwave replied:
Not sure how the cactus got in there, but leaving it because it is cute 😆

pixlwave reviewed (commented):
This is only a review on the FFI layer, but looks great!. Can't wait for it to land 🤩

![spacex-2-3537887872](https://github.com/user-attachments/assets/7a3715f3-4ebe-4805-9618-3c515f8900d7)

stefanceriu commented on crates/matrix-sdk-ui/src/spaces/room.rs:
```diff
@@ -0,0 +1,76 @@
+use ruma::{OwnedMxcUri, OwnedRoomAliasId, OwnedRoomId};
+
+#[derive(Debug)]
+pub struct SpaceServiceRoom {
```
Let's just keep it simple and just call it a SpaceRoom so it matches the spec `A space is defined by the [m.space room type](https://spec.matrix.org/latest/client-server-api/#types), making it known as a “space-room”.`

Hywan reviewed (changes requested):
I've reviewed up to the graph thingy. Is it possible to squash commits that are updating the same part of the code, they are mostly fixup at this point and it would simplifiy the review process.

I'm also wondering if using `Client::rooms_stream` could remove the need for some tasks, what do you think?

stefanceriu commented on crates/matrix-sdk-ui/src/spaces/mod.rs:
```diff
@@ -0,0 +1,148 @@
+            .collect()
+    }
+
+    pub async fn top_level_children_for(
```
This has since been replaced by `joined_spaces` which is also documented

stefanceriu commented on crates/matrix-sdk-ui/src/spaces/mod.rs:
I would expect `spaces` to grow and potentiall encompass more than just this initial service. Figured it's safer to have something that can group all space related functionality instead of narrowing down it's scope too soon.

stefanceriu commented on crates/matrix-sdk-ui/src/spaces/mod.rs:
```diff
@@ -0,0 +1,148 @@
+
+        println!("Top level children for space {}: {:?}", space_id, result.rooms);
+
+        Ok(vec![])
```
It is, with this being the initial commit. I'm happy to squash these if you think evolution history isn't relevant.

stefanceriu commented on crates/matrix-sdk-ui/src/spaces/room_list.rs:
```diff
@@ -0,0 +1,223 @@
+    pagination_state: SharedObservable<SpaceServiceRoomListPaginationState>,
+
+    rooms: SharedObservable<Vec<SpaceServiceRoom>>,
+}
```
With the APIs being very fluid and VectorDiffs being introduced todawards the end, I intentionall left documentation as one of the last steps. Handled in c6a25e99ab399139d716c76f156931bea3bbec05

stefanceriu commented on crates/matrix-sdk-ui/src/spaces/room_list.rs:
```diff
@@ -36,18 +39,60 @@ pub struct SpaceServiceRoomList {
+                    }
+                }
+            }
+        });
```
Talked about if offline: `rooms_stream` will trigger only when rooms get left or joined and not when any of their data changes.

stefanceriu commented on crates/matrix-sdk-ui/src/spaces/mod.rs:
```diff
@@ -0,0 +1,148 @@
+        self.client
+            .joined_rooms()
+            .into_iter()
+            .filter_map(|room| if room.is_space() { Some(room.room_id().to_owned()) } else { None })
```
Fair point. I think adding a `joined_space_rooms` method is the most straight forward solution, changing `RoomStateFilter` to do more than membership feels a bit too intrusive. One note though, it will return *any* joined space room, not the top level ones. We need the graph and what this service provides to compute those.

poljar commented on bindings/matrix-sdk-ffi/src/sync_service.rs:
```diff
@@ -90,6 +90,15 @@ impl SyncService {
             }
         })))
     }
+
+    /// Force expiring both sliding sync sessions.
+    ///
+    /// This ensures that the sync service is stopped before expiring both
+    /// sessions. It should be used sparingly, as it will cause a restart of
+    /// the sessions on the server as well.
+    pub async fn expire_sessions(&self) {
+        self.inner.expire_sessions().await;
+    }
```
Let's put this into a separate PR. Functionally-wise it's separate from the rest of the PR.

poljar commented on testing/matrix-sdk-test/src/event_factory.rs:
```diff
@@ -871,9 +873,11 @@ impl EventFactory {
         &self,
         creator_user_id: &UserId,
         room_version: RoomVersionId,
+        room_type: Option<RoomType>,
```
Instead of adding this argument here, we should add a new method here:

https://github.com/matrix-org/matrix-rust-sdk/blob/3aeface647183345973e310930fa037bc603db04/testing/matrix-sdk-test/src/event_factory.rs#L453-L465

This removes the need to modify *every* call site of this method which in turn makes the PR smaller. This change could have been a separate PR as well.

stefanceriu replied:
Oh this is neat, didn't realise it aws a thing! 👍

poljar commented on crates/matrix-sdk-ui/Cargo.toml:
```diff
@@ -67,6 +67,7 @@ assert-json-diff.workspace = true
 assert_matches.workspace = true
 assert_matches2.workspace = true
 eyeball-im-util.workspace = true
+futures-executor.workspace = true
```
What's this for?

poljar commented on crates/matrix-sdk-ui/src/spaces/mod.rs:
```diff
@@ -0,0 +1,510 @@
+/// use matrix_sdk_ui::spaces::SpaceService;
+/// use ruma::owned_room_id;
+///
+/// # futures_executor::block_on(async {
```
Oh, it's for this? Since the example is `no_run` you don't really need this, having the `async` block is enough, for example like here: https://github.com/matrix-org/matrix-rust-sdk/blob/3aeface647183345973e310930fa037bc603db04/crates/matrix-sdk-ui/src/sync_service.rs#L572-L603.

stefanceriu replied:
Ah sorry, this was supposed to work without `no_run` through that `futures-executor` but it ends up just failing miseraby with something along the lines of:

```
% cargo test spaces -p matrix-sdk-ui --doc

test crates/matrix-sdk-ui/src/spaces/mod.rs - spaces::SpaceService (line 60) ... FAILED

failures:

---- crates/matrix-sdk-ui/src/spaces/mod.rs - spaces::SpaceService (line 60) stdout ----
Test executable failed (exit status: 101).

stderr:

thread 'main' (14505151) panicked at /var/folders/xy/jh61h2x979g3r7c5rglgd_800000gn/T/rustdoctestbJlRWS/doctest_bundle_2024.rs:48:22:
not yet implemented
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

poljar replied:
We discussed this in chat. Using `no_run` is completely fine, we don't want our examples to try to reach real homeservers anyways.

poljar commented on testing/matrix-sdk-test/src/event_factory.rs:
```diff
@@ -987,6 +991,30 @@ impl EventFactory {
         self.event(CallNotifyEventContent::new(call_id, application, notify_type, mentions))
     }
 
+    /// Create a new `m.space.child` state event.
+    pub fn space_child(
+        &self,
+        parent: OwnedRoomId,
+        child: OwnedRoomId,
+    ) -> EventBuilder<SpaceChildEventContent> {
+        let mut event = self.event(SpaceChildEventContent::new(vec![]));
+        event.room = Some(parent);
+        event.state_key = Some(child.to_string());
+        event
+    }
+
+    /// Create a new `m.space.parent` state event.
+    pub fn space_parent(
+        &self,
+        parent: OwnedRoomId,
+        child: OwnedRoomId,
+    ) -> EventBuilder<SpaceParentEventContent> {
+        let mut event = self.event(SpaceParentEventContent::new(vec![]));
+        event.state_key = Some(parent.to_string());
+        event.room = Some(child);
+        event
+    }
+
```
Let us pack this into the same PR as the rest of the event factory changes.

poljar commented on crates/matrix-sdk-base/src/sync.rs:
```diff
@@ -90,13 +90,22 @@ impl RoomUpdates {
             .chain(self.invited.keys())
             .chain(self.knocked.keys())
     }
+
+    /// Returns whether or not this update contains any changes to the list
+    /// of invited, joined, knocked or left rooms.
+    pub fn is_empty(&self) -> bool {
+        self.invited.is_empty()
+            && self.joined.is_empty()
+            && self.knocked.is_empty()
+            && self.left.is_empty()
+    }
```
We can put this into a separate PR, those methods are useful on their own.

Don't forget a changelog entry.

poljar commented on crates/matrix-sdk/src/client/mod.rs:
```diff
@@ -1235,6 +1235,15 @@ impl Client {
             .collect()
     }
 
+    /// Returns the joined space rooms this client knows about.
+    pub fn joined_space_rooms(&self) -> Vec<Room> {
+        self.base_client()
+            .rooms_filtered(RoomStateFilter::JOINED)
+            .into_iter()
+            .flat_map(|room| room.is_space().then_some(Room::new(self.clone(), room)))
+            .collect()
+    }
+
```
Can be a separate PR, with a changelog.

poljar commented on crates/matrix-sdk/src/room/mod.rs:
```diff
@@ -1166,7 +1166,7 @@ impl Room {
         // https://spec.matrix.org/v1.8/client-server-api/#mspaceparent-relationships
 
-        // Get all m.room.parent events for this room
+        // Get all m.space.parent events for this room
```
These small comment fixes could also go into a separate PR.

stefanceriu commented:
Status update:
* I've extracted several small PRs from this one in #5592, #5593, #5594, #5595 and #5596
* with those merged onto main I went ahead and rebased this main PR on top of it and cleaned up the history

poljar reviewed (changes requested):
Alright, there are a couple of improvements that we should make, and a bunch of smaller nits, but overall this looks great.

Thank you again for splitting it up a bit, there's much less noise in the PR now.

stefanceriu commented on crates/matrix-sdk-ui/src/spaces/room_list.rs:
```diff
@@ -0,0 +1,406 @@
+}
+
+impl SpaceRoomList {
+    /// Creates a new `SpaceRoomList` for the given space identifier.
+    pub fn new(client: Client, parent_space_id: OwnedRoomId) -> Self {
```
It would be if I apply your suggestion?

poljar replied:
Oh, you mean return `None` if the space ID isn't a space? Or which suggestion? I didn't post a github code suggestion here, am I growing senile? :older_man:

stefanceriu replied:
Lol, nop, it's not you, it's me. I completely missread your original comment as wanting to return `SpaceRoomList?` from that existing constructor to which figured we might as well return a full result.

Yeah, the `SpaceRoom` is supposed to be a lightweight record that powers items in the rendered list, adding a client to them so we can build the RoomList might be a bit too much.

poljar replied:
Right, makes sense.

stefanceriu commented on crates/matrix-sdk-ui/src/spaces/room_list.rs:
```diff
@@ -0,0 +1,406 @@
+/// }
+/// // }
+///
+/// // spawn {
+/// while let Some(diffs) = room_stream.next().await {
+///     println!("Received room list update: {diffs:?}");
+/// }
+/// // }
```
Ups, sure thing 👍

poljar commented on crates/matrix-sdk-ui/src/spaces/mod.rs:
```diff
@@ -0,0 +1,491 @@
+//!
+//! It consists of 3 main components:
+//! - `SpaceService`: The main service for managing spaces. It
+//! - `SpaceGraph`: An utility that maps the `m.space.parent` and
+//!   `m.space.child` fields into a graph structure, removing cycles and
+//!   providing access to top level parents.
```
Yes :sweat_smile:

stefanceriu commented on crates/matrix-sdk-ui/src/spaces/mod.rs:
```diff
@@ -0,0 +1,491 @@
+            }
+        }
+
+        // Remove cycles from the graph. This is important because they are not
+        // enforced backend side.
+        graph.remove_cycles();
```
Yeah, let's see how this evolves, we might end up just replacing the whole graph component.

stefanceriu commented on crates/matrix-sdk-ui/src/spaces/graph.rs:
```diff
@@ -0,0 +1,217 @@
+    /// Adds a directed edge from `parent_id` to `child_id`, creating nodes if
+    /// they do not already exist in the graph.
+    pub fn add_edge(&mut self, parent_id: OwnedRoomId, child_id: OwnedRoomId) {
+        self.nodes.entry(parent_id.clone()).or_insert(SpaceGraphNode::new(parent_id.clone()));
+
+        self.nodes.entry(child_id.clone()).or_insert(SpaceGraphNode::new(child_id.clone()));
+
+        self.nodes.get_mut(&parent_id).unwrap().children.insert(child_id.clone());
+        self.nodes.get_mut(&child_id).unwrap().parents.insert(parent_id);
```
Definitely, nice one! 👍

stefanceriu commented on crates/matrix-sdk-ui/src/spaces/room_list.rs:
```diff
@@ -0,0 +1,406 @@
+        let mut request = get_hierarchy::v1::Request::new(self.parent_space_id.clone());
+        request.max_depth = Some(uint!(1)); // We only want the immediate children of the space
+
+        if let PaginationToken::HasMore(ref token) = *self.token.lock() {
+            request.from = Some(token.clone());
+        }
```
TIL and sounds like a good idea to me 👍

stefanceriu commented on crates/matrix-sdk-ui/src/spaces/mod.rs:
```diff
@@ -0,0 +1,491 @@
+            let joined_spaces = Arc::clone(&self.joined_spaces);
+            let all_room_updates_receiver = self.client.subscribe_to_all_room_updates();
+
+            *self.room_update_handle.lock() = Some(AbortOnDrop::new(spawn(async move {
```
Nice catch, thanks!

poljar reviewed (changes requested):
Alright, almost good to go.

There's one thing that I miscommunication a bit and even gave you an incorrect suggestion, sorry about that.

Let me know if you are out of time and want me to fix this myself.

stefanceriu commented on crates/matrix-sdk-ui/src/spaces/room_list.rs:
```diff
@@ -201,7 +201,7 @@ impl SpaceRoomList {
         }
 
-        self.pagination_state.set(SpaceRoomListPaginationState::Loading);
+        self.pagination_state.set_if_not_eq(SpaceRoomListPaginationState::Loading);
```
Talked about if offline, makes sense, made the change 👍

poljar replied:
“Offline” :sweat_smile:

poljar reviewed (approved):
Nice, I think this is good to go.

Congrats, now you have your own personal subsystem in the SDK that you have the privilege to maintain *forever* :scream:
